### PR TITLE
Fix Swift 4 ambiguity

### PR DESCRIPTION
**Pull request by tkohout**
Latest Xcode 9 beta 3 shows `ambiguous use of curry` when the swift 4 version flag is set for the project.
 
screen shot 2017-07-18 at 9 34 16 pm

This is likely related to changes in compiler implementation in swift 4 -
 see more here: https://github.com/Swinject/SwinjectAutoregistration/issues/29

The solution is to disambiguate by adding parentheses to functions with more than one parameter. As far as I tested this, it should be backward compatible with earlier versions of swift.

I would wait before merging this to see whether it is a bug of compiler in the beta or a real change.

### CODE REVIEW

File: bin/generate [hunk 1]
```diff
@@ -39,11 +43,11 @@ func curryDefinitionGenerator(arguments: Int) -> String {
   let lowerFunctionArguments = inputParameters.map { "\($0.lowercased())" } // ["a", "b", "c"]
   let returnType = genericParameters.last! // "D"
 
-  let functionArguments = "(\(commaConcat(inputParameters)))" // "(A, B, C)"
+  let functionArguments = extraParentheses("(\(commaConcat(inputParameters)))", when: inputParameters.count > 1) // "(A, B, C)"
```
Comment (tonyd256) [inline bin/generate:46]: what happens if there is only one argument? does it error out? could we skip this condition?

Comment (tkohout) [reply bin/generate:46]: yep, you get something like `Cannot convert value of type '(String) -> Service1' to expected argument type '((_)) -> _'`. There can't be extra parentheses for a single parameter.

**Comment by Andrewpk** [comment]
Is there anything holding this up from getting merged?

**Comment by tonyd256** [comment]
Not sure who's maintaining this anymore ... @sharplet ?

**Comment by gfontenot** [comment]
It's surprising to me that this works at all, beyond that it's needed. My understanding was that Swift moved _away_ from letting you pass tuples to functions that take multiple arguments, and yet this seems to be exactly what this is doing?

**Comment by wongzigii** [comment]
Bump

**Comment by johnnysay** [comment]
Any clue about when this will be merged?

**Comment by gfontenot** [comment]
I'll get this into a release today.

File: Curry.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -454,7 +454,7 @@
 				ONLY_ACTIVE_ARCH = YES;
 				SDKROOT = iphoneos;
 				SWIFT_OPTIMIZATION_LEVEL = "-Onone";
-				SWIFT_VERSION = 3.0;
+				SWIFT_VERSION = 4.0;
```
Comment (gfontenot) [inline Curry.xcodeproj/project.pbxproj:457]: Does this mean we should do a major version bump? @sharplet I feel like you know more about this compatibility than I do.

Comment (sharplet) [reply Curry.xcodeproj/project.pbxproj:457]: Yeah I'm pretty sure this would break compilation on Xcode 8.x

Comment (gfontenot) [reply Curry.xcodeproj/project.pbxproj:457]: 😞 _sigh_ that's what I was afraid of. Curry 4.0 here we come!

**Comment by gfontenot** [comment]
Curry 4.0 is now released, which includes this change and so should fix compatibility with Swift 4.

**Comment by gfontenot** [comment]
Sorry for the delay on this btw, and thank you for your PR and your patience.

**Comment by johnnysay** [comment]
Thank you @gfontenot for your fast reactivity!
However when running `pod install` / `pod update` I still have the version `3.0.0`.
I have also noticed that in https://github.com/thoughtbot/Curry/blob/master/Curry.podspec the version is still `3.0.0`.

**Comment by gfontenot** [comment]
d'oh! Thanks. I'll get that pushed.

**Pull request merged**